### PR TITLE
Retry on database error to reduce number of test failures

### PR DESCRIPTION
**Pull request by mfeurer**
#### What does this PR implement/fix? Explain your changes.

This PR adds a retry mechanism to api calls in case it detects a database error.

#### How should this PR be tested?

New unit test. Also, an overloaded test server.

### CODE REVIEW

File: openml/_api_calls.py [hunk 1]
```diff
@@ -198,17 +188,42 @@ def __send_request(
                     response = session.post(url, data=data, files=files)
                 else:
                     raise NotImplementedError()
-                break
+                pass
```
Comment (PGijsbers) [inline openml/_api_calls.py:191]: `pass` has no function here

File: openml/_api_calls.py [hunk 2]
```diff
@@ -198,17 +188,42 @@ def __send_request(
...
+            except OpenMLServerException as e:
+                # Do retries in case of a database connection error
+                if e.code == 107:
+                    if i == n_retries:
+                        raise e
+                    else:
+                        time.sleep(i * 0.3)
+                        continue
+                else:
+                    raise e
```
Comment (PGijsbers) [inline openml/_api_calls.py:210]: Maybe simplify the flow by calling `check_response` where you currently have the `pass` statement. You can then catch the `OpenMLServerException` in the same `except` clause and use shared logic on whether or not an error should be raised or a new connection attempt should be made.

**Review by PGijsbers — CHANGES_REQUESTED** [review]
I think it makes sense to treat a connection error the same as a `server is busy` is error. I have a minor suggestion for making sure the code reflects that. Other than that I agree with the PR.

File: openml/_api_calls.py (hunk 2 above)
Comment (mfeurer) [reply openml/_api_calls.py:210]: You mean like this?
```
            try:
                if request_method == "get":
                    response = session.get(url, params=data)
                elif request_method == "delete":
                    response = session.delete(url, params=data)
                elif request_method == "post":
                    response = session.post(url, data=data, files=files)
                else:
                    raise NotImplementedError()
                __check_response(response=response, url=url, file_elements=files)
                break
            except (
                requests.exceptions.ConnectionError, 
                requests.exceptions.SSLError,
                OpenMLServerException,
            ) as e:
                if isinstance(e, OpenMLServerException) and e.code != 107:
                    # 107 is a database connection error - only then do retries
                    raise e
                if i == n_retries:
                    raise e
                else:
                    time.sleep(0.1 * i)
                    continue
```
Or a bit more verbose like that:
```
            try:
                if request_method == "get":
                    response = session.get(url, params=data)
                elif request_method == "delete":
                    response = session.delete(url, params=data)
                elif request_method == "post":
                    response = session.post(url, data=data, files=files)
                else:
                    raise NotImplementedError()
                __check_response(response=response, url=url, file_elements=files)
                break
            except (
                requests.exceptions.ConnectionError,
                requests.exceptions.SSLError,
            ) as e:
                if i == n_retries:
                    raise e
                else:
                    time.sleep(0.1 * i)
                    continue
            except OpenMLServerException as e:
                # 107 is a database connection error - only then do retries
                if e.code != 107:
                    raise
                if i == n_retries:
                    raise e
                else:
                    time.sleep(0.3 * i)
                    continue
```
In any case, merging the two try/excepts is an improvement to the code.

Comment (PGijsbers) [reply openml/_api_calls.py:210]: Fine with either, but I think I prefer the first one (with a `wait_time` variable if you want to preserve the difference in sleep time).

**Review by PGijsbers — APPROVED** [review]
(let's see some green from CI first, but otherwise 👍 )

**Comment by joaquinvanschoren** [comment]
@all-contributors please add @mfeurer for code

**Pull request merged**